smart-connector/src/main/java/eu/knowledge/engine/smartconnector/impl/ReasonerProcessor.java

```diff
@@ -151,15 +160,38 @@ public void planAskInteraction(MyKnowledgeInteractionInfo aAKI) {
 		}
 	}
 
+	/**
+	 * In this ask operation that smart connector should do the following:
+	 * 1: first, make a plan for executing the ask using the planAsk method in this smart connector
+	 *    this will return an object of type AskPlan.
+	 * 2: second, execute the plan on the knowledge network to potentially get bindings
+	 *    for the pattern in the ask. 
+	 * 3: third, as part of the execution also find knowledge gaps in the reasoner plan when the
+	 *    resulting binding set is empty. This will result in an object of type AskResult that
+	 *    contains the resulting bindings, exchange info and 
+	 *    optionally the reasoner plan plus knowledge gaps.
+	 * 
+	 * This can lead to the following situations:
+	 * 1: a plan, a non-empty binding set and no gaps => ask has a result
+	 * 2: a plan, an empty binding set and no gaps => ask has an empty result
+	 * 3: a plan, an empty binding set with gaps => ask has no result and gaps are found
+	 * 
+	 */
 	@Override
 	public CompletableFuture<AskResult> executeAskInteraction(BindingSet someBindings) {
 
 		this.finalBindingSetFuture = new CompletableFuture<eu.knowledge.engine.reasoner.api.BindingSet>();
 //		this.reasonerPlan.optimize();
 		continueReasoningBackward(translateBindingSetTo(someBindings));
-
+		
 		return this.finalBindingSetFuture.thenApply((bs) -> {
-			return new AskResult(translateBindingSetFrom(bs), this.askExchangeInfos, this.reasonerPlan);
+			//this.knowledgeGaps = new HashSet<>();
+			this.knowledgeGaps = new KnowledgeGapSet();
+			// if the binding set is empty, check for knowledge gaps
+			if (bs.isEmpty()) {
```

Comment:
We might consider making the inclusion of knowledge gaps configurable, because I can imagine not all scenarios are interested in those and they might add some performance penalty. We can, however, also do that at a later stage when improving the configuration of the knowledge engine (#512) is done.

Reply:
Ok...for now, the calculation of knowledge gaps is only executed when there is an empty binding. This seems to be THE situation when this needs to happen. But, of course it might be better to let the user/caller decide whether or not to calculate knowledge gaps.